### DISCUSSION TITLE
Get access to DataFetchingEnvironment from KotlinDataLoader

### DISCUSSION THREAD

**Discussion by JajaComp** [ExpediaGroup/graphql-kotlin#1145] [Q&A] [answered]
Hi! Can i get access to DataFetchingEnvironment from KotlinDataLoader? Or how i can add arguments to subRequests?

**Comment by smyrick** [comment, marked as answer]
`KotlinDataLoader` is an abstraction on graphql-java DataLoader that the spring server library will autowire for you.

You can always manually create and load data loaders where you can give it info in the `DataFetchingEnvironment`. Infact during execution, you have to use the `DataFetchingEnvironment` to get data loaders in the first place.

See the example app:

https://github.com/ExpediaGroup/graphql-kotlin/blob/2a498e85e0cd2104dfcc3ccb7a3d73ca225fc92a/examples/server/spring-server/src/main/kotlin/com/expediagroup/graphql/examples/server/spring/query/DataLoaderQuery.kt#L47-L57

You can get more info from the data loader docs directly: https://github.com/graphql-java/java-dataloader

**Reply by JajaComp** [reply]
Ok. Thank you. Also can i exclude field from response if sub field meets some condition? Example ->
```
query {
  person(where: {place: {city: {eq: "New York"}}}) {
    name
    place {
      city
    }
  }
}
```

**Reply by smyrick** [reply]
You can't change the client query. If they request a field you must return it following the GraphQL spec. 

Clients could modify their own query with the `@include` or `@skip` directives, but this is controlled by the client and you can't modify that behaviour server side or have it conditional based on the response

```graphql
query GetPerson($skipName: Boolean) {
  person 
    name @skip(if: $skipName)
    place {
      city
    }
  }
}
```

You could have the field be nullable so on the backend you could return or not return the field based of some logic though

```kotlin
fun person(city: String?): List<Person>? = findPeopleByCity(city)
```

**Reply by JajaComp** [reply]
I want to implement as done in Hasura - there you can put where conditions on nested entities. For example, in such a structure, I want to get users who live in new york:
```
{
   person (where: {city: {name: {eq: "New york"}}}) {
     name
     city {
       name
     }
   }
}
```

Is it out of spec?
I must do other request with custom "where" and add custom logic for get data from db?

**Reply by smyrick** [reply]
That is not out of spec, that is just input to your `person` field. As above I made it just an single `String` argument but you could do this if you wanted. 

```kotlin
fun person(where: WhereInput): Person? = findPeople(where)

class WhereInput(val city: CityInput)

class CityInput(val name: StringInput)

class StringInput(val eq: String)
```


In my opinion this input seems a bit verbose from an API perspective, and is usually something you want to abstract away from API users from your DB, so I would personally just use a single `city: String?` argument but you can use whatever you want

**Reply by JajaComp** [reply]
I am having a problem understanding architecture. With parameters, everything is simple, but how to filter out unnecessary data? At first I thought that it was possible to do this using DataLoader, but as I understand it, in this case it is necessary to refuse and use a single query and get all the people from the database myself, then manually get the cities of these users (Or use join), manually fill in the structure the returned data, through the Context, prevent the use of DataLoader Or is there a better way?

**Reply by smyrick** [reply]
You could fetch all people from your DB if you wanted to do extra logic that a DB search can't handle and then do the filter yourself, that would be an option but maybe not the best

```kotlin
fun person(city: String?): List<Person> {
  val allPeople: List<People> = getAllPeopleFromDB()
  
  return if (city == null) {
    allPeople
  } else {
    allPeople.filter { person -> person.city == city }
  }
}
```

But it would probably be better to form a DB query based off the API input then run that. No need for data loaders or context

```kotlin
const val query = "SELECT * FROM people WHERE city = ?"

fun person(city: String?): List<Person> {
  return if (city == null) {
    getAllPeopleFromDB()
  } else {    
    val dbQuery = buildQuery(query, city)
    return runQuery(dbQuery)
  }
}
```

The point is that all this is hidden from API users and all they do is this

```graphql
query {
  person(city = "New York") {
   name
  }
}

**Reply by smyrick** [reply]
All of this is separate from GraphQL. This is part of schema design and the whole point of having an API on top of your DB is to hide implementation details.

Try to think of what use cases your API clients are doing and design the schema to serve them best. Then it is the server side job to handle all the technical details. You probably don't want to just expose everything your DB can do directly

**Reply by JajaComp** [reply]
And i can't use custom logic in DataLoader for exclude useless data like in my sample? It is important for me to understand exactly the possible logic in subqueries.

**Reply by smyrick** [reply]
To implement the schema you shared above you should have a top level kotlin function called `person`

```kotlin
fun person(input: WhatEverYouWant): List<Person> { 
  TODO("implement me")
}
```

In this function you need to some how get the list of people. You can do that by calling your database, calling another API, generate it randomly or any other option. If you want to use DataLoader, it is just a cache on top of your call and it returns the objects.

Once you have your list of people you can write addition server-side code to modify results anyway you want based on headers, query input, or the response from your calls.

It sounds like you may be misinterpreting how GraphQL in general works and where this library and the schema generator fits in to your tech stack. I recommend you first check out some of our example videos: https://expediagroup.github.io/graphql-kotlin/docs/blogs-and-videos
